### ISSUE TITLE
📎 Remove compiler diagnostics for deleted nodes

### ISSUE THREAD

**Issue by sebmck** [rome/tools#803]
Right now we stack them up. Which is bad and results in excessive deduping.

**Comment by VictorHom** [comment]
@sebmck is this something I can help with?

**Comment by sebmck** [comment]
@VictorHom If you'd like! I'm not quite sure how this would work. Basically diagnostics can refer to nodes that have already been deleted.

[`packages/@romefrontend/compiler/methods/reduce.ts`](https://github.com/romefrontend/rome/blob/main/packages/%40romefrontend/compiler/methods/reduce.ts) is where we handle AST traversal. We can handle the simple case of "deleted node <-> diagnostic" by maintaining a map. Where it gets significantly more complicated is when it's a parent that gets deleted instead.

Could be solved with some more complicated tracking by keeping a track of parents when registering a diagnostic, and adding hooks for AST creation/deletion. Might be tricky from a performance perspective though. Although a performance penalty only when there are diagnostics in a file might be ok (which is what we would do).

We might also need to move the diagnostic creation logic to the `Path` as it's where we have ancestry/parent information.

**Comment by jamiebuilds** [comment]
Hi! Thank you for opening this issue

We recently made the decision to [migrate the Rome codebase to Rust](https://rome.tools/blog/2021/09/21/rome-will-be-rewritten-in-rust) and making some significant changes to its architecture. We've also decided not to release the previous version written in TypeScript.

We're going through and closing pull requests and issues that were specific to the TypeScript codebase, including this one. Thank you for taking the time to open it.

Thanks again!
